### DISCUSSION TITLE
How to define cyclic dependencies?

### DISCUSSION THREAD

**Discussion by ceving** [vanjs-org/van#217] [Q&A]
`xx` should be the square of `x` and `x` should be the square root of `xx`.

Example:

~~~js
window.x  = van.state(() => 0);
window.xx = van.derive(() => window.x.val * window.x.val);
window.x  = van.derive(() => Math.sqrt(window.xx.val));
~~~

This works but only one-way:

~~~js
xx.val = 4 ;
x.val // => 2
x.val = 3;
xx.val // => 4
~~~

**Comment by Tao-VanJS** [comment]
Try this: https://codepen.io/taoxin/pen/KKEpyQR?

**Reply by ceving** [reply]
That creates an insane amount of function calls for just one click. All the "update helper" lines are generated by just one click reducing b from 4 to 3.
![Screenshot 2023-12-28 at 22-40-47 wvOavEp](https://github.com/vanjs-org/van/assets/795748/517890bf-9a3a-4ec0-a00c-c778afecef03)

**Reply by Tao-VanJS** [reply]
I see your point. Will try to see if there is any better solution.

**Reply by Tao-VanJS** [reply]
Hi @ceving,

I modified the code to make it work with the help of a helper function `withCustomValSetter`. You can check out the new version: https://codepen.io/taoxin/pen/wvOavEp.

I don't have a complete understanding of why the problem occured in the old version in the first place. Will spend some time investigating to see if I can make it better.

**Reply by Tao-VanJS** [reply]
Hi @ceving,

With some troubleshooting, I found the root cause of problem which calls the derivation function many times. Basically, I need to convert the `string` value into `number` value (which can be achieved with `parseInt`) before setting the states for `a` and `b`. Otherwise, it will mess up the equality check of states by comparing a `string` value with a `number` value. After the fix, you can find the derivation function is called exactly once whenever the state changes.

See the new version of the implementation: https://codepen.io/taoxin/pen/LYapVXJ.

Hope it helps :-)

**Reply by ceving** [reply]
This does not detect cycles. If you enter `a.val = 4` in the console, you will see, that this modifies first `b` and then again `a`. I think it should not be allowed that a change of `a` triggers a change of `a`. This leads to an endless recursion as in the following example:
~~~js
// Pythagoras

const {p, div, input} = van.tags;

var edge = { a: van.state(3), b: van.state(4), c: van.state(5) };

Object.keys(edge).forEach(name =>
  van.derive(() => console.log(`${name}: ${edge[name].val}`)));

van.derive(() => edge.a.val = Math.sqrt(Math.pow(edge.c.val, 2) - Math.pow(edge.b.val, 2)));
van.derive(() => edge.b.val = Math.sqrt(Math.pow(edge.c.val, 2) - Math.pow(edge.a.val, 2)));
van.derive(() => edge.c.val = Math.sqrt(Math.pow(edge.a.val, 2) + Math.pow(edge.b.val, 2)));

function number(n)
{ return div(`${n}: `,
	     input({type: 'number', value: edge[n],
		    oninput: e => edge[n].val = parseInt(e.target.value)})); }

van.add(document.body, p(Object.keys(edge).map(number)));
~~~

**Reply by vrcxz** [reply]
> I think it should not be allowed that a change of a triggers a change of a. 

But isn't van.state() core function is to basically trigger the functions in its listener whenever it is changed?

And purposely triggering it to change inside a function that calls it when changed would undoubtedly not end up good.

As for your example, I think this should do it:

```js
// Pythagoras
const {p, div, input} = van.tags;

var edge = { a: van.state(3), b: van.state(4), c: van.state(5) };
var ui = {}

Object.keys(edge).forEach(name =>
  van.derive(() => console.log(`${name}: ${edge[name].val}`)));

ui.a = van.derive(() => Math.sqrt(Math.pow(edge.c.val, 2) - Math.pow(edge.b.val, 2)));
ui.b = van.derive(() => Math.sqrt(Math.pow(edge.c.val, 2) - Math.pow(edge.a.val, 2)));
ui.c = van.derive(() => Math.sqrt(Math.pow(edge.a.val, 2) + Math.pow(edge.b.val, 2)));

function number(n)
{ return div(`${n}: `,
	     input({type: 'number', value: ui[n],
		    oninput: e => edge[n].val = parseInt(e.target.value)})); }

van.add(document.body, p(Object.keys(edge).map(number)));

```

---

~~I think this one also works, but I'm not sure if this is a good practice:~~
This one actually don't seem to work. The previous states still seem to exist and are still reactive so it would end up in a recursion. I think.

```js
// Pythagoras
const {p, div, input} = van.tags;

var edge = { a: van.state(3), b: van.state(4), c: van.state(5) };

Object.keys(edge).forEach(name =>
  van.derive(() => console.log(`${name}: ${edge[name].val}`)));

edge.a = van.derive(() => Math.sqrt(Math.pow(edge.c.val, 2) - Math.pow(edge.b.val, 2)));
edge.b = van.derive(() => Math.sqrt(Math.pow(edge.c.val, 2) - Math.pow(edge.a.val, 2)));
edge.c = van.derive(() => Math.sqrt(Math.pow(edge.a.val, 2) + Math.pow(edge.b.val, 2)));

function number(n)
{ return div(`${n}: `,
	     input({type: 'number', value: edge[n],
		    oninput: e => edge[n].val = parseInt(e.target.value)})); }

van.add(document.body, p(Object.keys(edge).map(number)));
```

**Reply by Tao-VanJS** [reply]
My take is: circular dependencies among states are inherently error-prone. This is not something I would recommend. Oftentimes the code will be easier to maintain with alternative solutions. Even the solution like https://codepen.io/taoxin/pen/wvOavEp is less error prone than direct circular dependencies among states.

For the example like https://codepen.io/taoxin/pen/LYapVXJ to work, we need to guarantee that: whenever `b` changes (driven by a UI event), it will trigger `a` to change which will lead the derivation of `b`, but this time, the derivation of `b` must be EXACTLY the same as its prior value (post UI-driven change), which effectively ends the recursive derivation. The same needs to be true whenever `a` changes as well. Failing to guarantee that means bugs in your app.

As an unopinionated framework, VanJS doesn't want to ban a certain feature or a certain coding style if it's possible to build apps this way. That said, we have to acknowledge that some features might be harder to use correctly and oftentimes there could be alternatives that can serve your need in an easier way. For the case of circular dependencies among states, you have to know exactly what you're doing to use it correctly.

**Reply by ceving** [reply]
It is necessary to allow a state to modify itself, if you want to allow recursion. But recursion normally requires, that you can define an arbitrary termination condition. VanJS seems to allow recursion but only with a very specific termination condition: $y(x-1) = y(x)$. On the other hand it would be possible to disallow recursion at all: an impact of a state change can not change the same state. This would makes the Pythagoras example pretty simple. I am wondering what is the benefit of allowing this somehow crippled kind of recursion? Which UI use-case would need it?

**Reply by Tao-VanJS** [reply]
In general, a state derivation represents an invariant. For instance:

```js
const b = van.derive(() => 10 - a.val)
```

represents the invariant of $b = 10 - a$.

It would be possible to define derivations in VanJS that triggered from both directions. But these derivations must represent invariants that are consistent with each other. For instance, in the example: https://codepen.io/taoxin/pen/LYapVXJ, the invariant $b = 10 - a$ and $a = 10 - b$ are consistent (as they are both equivalent to $a + b = 10$). Failing to ensure the consistency among invariants will cause problems, one way or another. This is the extra care you have to take if you want to trigger derivations from both derivations.

Detecting derivation loop and stopping at the loop, even if it's possible (which I'm not quite sure), will not lead to desirable outcome, either. Let's say in a hypothetical example with inconsistent invariants: $b = 10 - a$, $a = 9 - b$. If some UI event changes `b`, which lead to the re-derivation of `a`, which can potentially lead to the re-derivation of `b`. If we stop the re-derivation of `b` at the point of the loop detection, we're essentially breaking the invariant $b = 10 - a$. Thus, the root cause of the problem is inconsistent invariants, no solution will provide a reasonable behavior if inconsistent invariants are introduced to the program.

**Reply by ceving** [reply]
Is it possible to customize equivalence? This will not work: $b = a^2, a = \sqrt{b}$. For the equivalence it might be sufficient to say *"the numbers are the same if they are the same up to the nth decimal place"*, but for any additional computation you typically do not want to throw away the remaining digits, because this would make the error even worse.

**Reply by Tao-VanJS** [reply]
Equality comparison for floating point numbers is always tricky. Thus it's probably not a great idea to pair it with another tricky feature -  cyclic derivations.
